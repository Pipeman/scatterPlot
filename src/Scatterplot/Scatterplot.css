.scatterplot-container {
  margin: 20px;
}