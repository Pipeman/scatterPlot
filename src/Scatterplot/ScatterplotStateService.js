export const initialState = {
  points: [],
  ranges: {
    xDomainRange:[0,0],
    yDomainRange:[0,0],
  },
};

export const setPoints = points => (state = initialState) => {    
    let data = points.map((point) => ({ x: parseFloat(point.x), y: parseFloat(point.y) }));
    let ranges = { ...state.ranges };
    data.forEach((point) => {
        if (point.x < ranges.xDomainRange[0]) {
          ranges.xDomainRange[0] = point.x;
        }
        if (point.x > ranges.xDomainRange[1]) {
          ranges.xDomainRange[1] = point.x;
        }
        if (point.y < ranges.yDomainRange[0]) {
          ranges.yDomainRange[0] = point.y;
        }
        if (point.y > ranges.yDomainRange[1]) {
          ranges.yDomainRange[1] = point.y;
        }
      }
    );

    return {
      points: data,
      ranges,
    }
}