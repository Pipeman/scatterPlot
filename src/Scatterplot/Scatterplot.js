import React, { Component } from 'react';
import { ScatterplotChart } from "react-easy-chart";
import './Scatterplot.css';
import { initialState, setPoints } from "./ScatterplotStateService"

class Scatterplot extends Component {
  constructor() {
    super();
    this.state = initialState;
  }

  componentDidMount() {
    const POINTS_URL = "https://scatterplotserver-itrvcanfvr.now.sh/points";
    let localPoints = localStorage.getItem("points");
    if (localPoints){
      this.setState(setPoints(JSON.parse(localPoints)));
    }
    fetch(POINTS_URL)
      .then(res => res.json())
      .then(({ points }) => {
        this.setState(setPoints(points));
        localStorage.setItem("points", JSON.stringify(points));
      });
  }

  componentWillUnmount() {
    localStorage.removeItem("points");
  }

  render() {
    let { points, ranges } = this.state;
    const RANGE_DIVIDER = 10;

    ranges.xDomainRange[1] += ranges.xDomainRange[1] / RANGE_DIVIDER;
    ranges.yDomainRange[1] += ranges.yDomainRange[1] / RANGE_DIVIDER;

    if(!points.length) {
      return <div>No Points at the moment. We might be fetching them.</div>;
    }

    return (
      <div className="scatterplot-container">
        <ScatterplotChart
            {...ranges}
            data={points}
            axes
            width={1000}
            height={500}
        />
      </div>
    );
  }
}

export default Scatterplot;
