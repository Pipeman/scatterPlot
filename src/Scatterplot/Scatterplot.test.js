import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Scatterplot from './Scatterplot';
import { initialState, setPoints } from './ScatterplotStateService';

const points = [{ x: 10, y: 10 }, { x: 34, y: 100 }];
const div = document.createElement('div');

beforeEach(() => {
  global.fetch = jest.fn(() => {
    return Promise.resolve({
      points,
      json() {
        return { points };
      }
    });
  });
});

it('renders without crashing', () => {
  const ScatterplotComponent = ReactDOM.render(<Scatterplot />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders No points message', () => {
  const ScatterplotComponent = renderer.create(<Scatterplot />);
  const ScatterplotComponentJSON = ScatterplotComponent.toJSON();
  expect(ScatterplotComponentJSON).toMatchSnapshot();
  expect(ScatterplotComponentJSON.children[0])
    .toBe("No Points at the moment. We might be fetching them.");
});

it('renders the chart', () => {
  const getLocalStorageSpy = jest.spyOn(localStorage, 'getItem')
    .mockImplementation(() => (JSON.stringify(points)));
  const ScatterplotComponent = renderer.create(<Scatterplot />);  
  
  expect(ScatterplotComponent.toJSON()).toMatchSnapshot();
});

it('state contains points and axes ranges', () => {
  const ScatterplotClass = new Scatterplot();
  expect(ScatterplotClass.state).toEqual(initialState);
});

it('fetch the points from the localStorage', () => {
  const getLocalStorageSpy = jest.spyOn(localStorage, 'getItem')
    .mockImplementation(() => (JSON.stringify(points)));
  const ScatterplotComponent = ReactDOM.render(<Scatterplot />, div);
  
  expect(getLocalStorageSpy).toHaveBeenCalled();
  expect(ScatterplotComponent.state.points).toEqual(points);
  ReactDOM.unmountComponentAtNode(div);
});

it('fetch the points from the url', done => {
  const getLocalStorageSpy = jest.spyOn(localStorage, 'getItem')
    .mockImplementation(() => null);
  const setLocalStorageSpy = jest.spyOn(localStorage, 'setItem')
    .mockImplementation(() => ({}));
  const ScatterplotComponent = ReactDOM.render(<Scatterplot />, div);
  
  expect(fetch).toBeCalledWith("https://scatterplotserver-itrvcanfvr.now.sh/points");
  setImmediate(() => {
    expect(ScatterplotComponent.state.points).toEqual(points);
    ReactDOM.unmountComponentAtNode(div);
    done();
  });
});

it('clears the local storage when the component unmounts', () => {
  const ScatterplotComponent = ReactDOM.render(<Scatterplot />, div);
  ReactDOM.unmountComponentAtNode(div);
  expect(localStorage.removeItem).toBeCalled();
});
