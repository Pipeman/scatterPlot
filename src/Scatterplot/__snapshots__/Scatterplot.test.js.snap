// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders No points message 1`] = `
<div>
  No Points at the moment. We might be fetching them.
</div>
`;

exports[`renders the chart 1`] = `
<div
  className="scatterplot-container"
>
  <div
    className="scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .pie-chart-slice{stroke: #fff;
        stroke-width: 1;
        opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .pie-chart-slice:hover{opacity: 0.8;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .pie-chart-label{font-family: sans-serif;
        font-size: 12px;
        text-anchor: middle;
        fill: #000;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .bar{fill: blue;
        transition: x 0.35s ease-in, y 0.35s ease-in, height 0.5s ease-in, width 0.5s ease-in;
        opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .bar:hover{opacity: 0.8;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .line{fill: none;
        stroke-width: 1.5;
        opacity: 0.7;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .line:hover{opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .area{opacity: 0.7;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .area:hover{opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .dot{stroke-width: 0;
        opacity: 0.85;
        transition: cx 0.35s ease-in, cy 0.35s ease-in, r 0.5s ease-in;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .dot:hover{opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 circle.data-point{r: 4px;
        opacity: 0.7;
        transition: cx 0.35s ease-in, cy 0.35s ease-in;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 circle.data-point:hover{r: 6px;
        opacity: 1;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 circle.tick-circle{r: 2px;
        fill: lightgrey;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .x circle.tick-circle{cy: 8px;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .axis{font-family: dobra-light,Arial,sans-serif;
        font-size: 9px;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .axis .label{font: 14px arial;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .axis path,.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .axis line{fill: none;
        stroke-width: 1;
        shape-rendering: crispEdges;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 x.axis path{display: none;
        stroke: lightgrey;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .tick line{stroke: lightgrey;
        stroke-width: 1;
        opacity: 0.7;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .x circle.tick-circle{fill: lightgrey;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .y circle.tick-circle{cx: -8px;
        fill: lightgrey;}.scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530 .y.axis line{display: none;
        stroke: lightgrey;}",
        }
      }
    />
    <svg
      height={563}
      style={Object {}}
      width={1072}
    >
      <g
        style={Object {}}
        transform="translate(48, 24)"
      >
        <g
          className="x axis"
          id="scatterplot-x-axis-5d14b9d78cd38f1509276ee3690d69858c851530"
          style={Object {}}
          transform="translate(0, 442)"
        />
        <g
          className="y axis"
          id="scatterplot-y-axis-5d14b9d78cd38f1509276ee3690d69858c851530"
          style={Object {}}
          transform="translate(0, 0)"
        />
        <g
          id="scatterplot-chart-5d14b9d78cd38f1509276ee3690d69858c851530"
          style={Object {}}
        />
      </g>
    </svg>
  </div>
</div>
`;
